pyro/params/param_store.py

```diff
@@ -223,19 +228,29 @@ def get_param(self, name, init_tensor=None, tags="default"):
 
             # a function
             if callable(init_tensor):
-                self._params[name] = init_tensor()
-            else:
-                # from the memory passed in
-                self._params[name] = init_tensor
+                init_tensor = init_tensor()
+
+            # store the unconstrained value and constraint
+            with torch.no_grad():
+                unconstrained_param = transform_to(constraint).inv(init_tensor)
```

Comment:
Is it worth skipping this for `constraints.real`? I suppose it should be fine unless there is some non-negligible cost to using the identity transform. It might be good to do a perf comparison (we can just use one of the tests that you modified, but we can do it separately).

Reply:
> It might also be good to do a perf comparison

Yeah, we should definitely add one, and there's a clear perf regression since the transform is applied every time `get_param` is called.  This could be avoided with a bit more bookkeeping, but we thought we'd address that in a followup PR.

Reply:
`transform_to(constraints.real).inv` simply returns its argument and is very cheap.